src/System.Linq/tests/ToArrayToListToDictionaryTests.cs

```diff
@@ -0,0 +1,469 @@
+        }
+
+
+        [Fact]
+        public void ToArray_WorkWithEmptyCollection()
+        {
+            RunToArrayOnAllCollectionTypes(new int[0],
+                resultArray =>
+                {
+                    Assert.NotNull(resultArray);
+                    Assert.Equal(0, resultArray.Length);
+                });
+        }
+
+        [Fact]
+        public void ToArray_ProduceCorrectArray()
+        {
+            int[] sourceArray = new int[] { 1, 2, 3, 4, 5, 6, 7 };
+            RunToArrayOnAllCollectionTypes(sourceArray,
+                resultArray =>
+                {
+                    Assert.Equal(sourceArray.Length, resultArray.Length);
+                    Assert.Equal(sourceArray, resultArray);
+                });
+
+            string[] sourceStringArray = new string[] { "1", "2", "3", "4", "5", "6", "7", "8" };
+            RunToArrayOnAllCollectionTypes(sourceStringArray,
+                resultStringArray =>
+                {
+                    Assert.Equal(sourceStringArray.Length, resultStringArray.Length);
+                    for (int i = 0; i < sourceStringArray.Length; i++)
+                        Assert.Same(sourceStringArray[i], resultStringArray[i]);
+                });
+        }
+
+
+        [Fact]
+        public void ToArray_TouchCountWithICollection()
+        {
+            CollectionTest<int> source = new CollectionTest<int>(new int[] { 1, 2, 3, 4 });
+            var resultArray = source.ToArray();
+
+            Assert.Equal(source, resultArray);
+            Assert.Equal(1, source.CountTouched);
+        }
+
+
+        [Fact]
+        public void ToArray_ThrowArgumentNullExceptionWhenSourceIsNull()
+        {
+            int[] source = null;
+            Assert.Throws<ArgumentNullException>(() => source.ToArray());
+        }
+
+
+        // Later this behaviour can be changed
+        [Fact]
+        public void ToArray_UseCopyToWithICollection()
+        {
+            CollectionTest<int> source = new CollectionTest<int>(new int[] { 1, 2, 3, 4 });
+            var resultArray = source.ToArray();
+
+            Assert.Equal(source, resultArray);
+            Assert.Equal(1, source.CopyToTouched);
+        }
+
+
+        // Currently not passed
+        //[Fact]
+        public void ToArray_WorkWhenCountChangedAsynchronously()
+        {
+            AsyncCollectionTest source = new AsyncCollectionTest(new int[] { 1, 2, 3, 4 });
+            var resultArray = source.ToArray();
+
+            Assert.True(resultArray.Length >= 4);
+            Assert.Equal(1, resultArray[0]);
+            Assert.Equal(2, resultArray[0]);
+            Assert.Equal(3, resultArray[0]);
+            Assert.Equal(4, resultArray[0]);
+        }
+
+
+        //[Fact]
+        public void ToArray_FailOnExtremelyLargeCollection()
+        {
+            LargeSequenceTest largeSeq = new LargeSequenceTest();
+            Assert.ThrowsAny<OverflowException>(() => { largeSeq.ToArray(); });
```

Comment:
This may be the case on 32-bit machines when there's no continuous block of 1Gb virtual memory. It is unlikely that such situation can occur during test running. But it is not a problem to handle it too.